usl_pipeline/usl_lib/usl_lib/chunkers/polygon_chunkers.py

```diff
@@ -0,0 +1,136 @@
+
+    chunk_col_count = min(chunk_size, global_col_count - x_chunk_index * chunk_size)
+    chunk_row_count = min(chunk_size, global_row_count - y_chunk_index * chunk_size)
+    chunk_max_x = chunk_min_x + chunk_col_count * cell_size
+    chunk_min_y = chunk_max_y - chunk_row_count * cell_size
+    return chunk_min_x, chunk_min_y, chunk_max_x, chunk_max_y
+
+
+def _bbox_intersection(
+    bbox1: Tuple[float, float, float, float], bbox2: Tuple[float, float, float, float]
+) -> bool:
+    """Checks if two bounding boxes have intersection.
+
+    Each bounding box is represented by a tuple (min-x, min-y, max-x, max-y).
+
+    Args:
+        bbox1: Bounding box 1.
+        bbox2: Bounding box 2.
+
+    Returns:
+        Indicator of the intersection of bounding boxes.
+    """
+    return max(bbox1[0], bbox2[0]) <= min(bbox1[2], bbox2[2]) and max(
+        bbox1[1], bbox2[1]
+    ) <= min(bbox1[3], bbox2[3])
+
+
+def split_polygons_into_chunks(
+    elevation_header: geo_data.ElevationHeader,
+    chunk_size: int,
+    polygon_masks: Iterable[Tuple[geometry.Polygon, int]],
+    output_dir_path: str | pathlib.Path,
+    chunk_file_name_pattern: str = "chunk_{y}_{x}",
+    support_mask_values=False,
+) -> list[chunkers_data.ChunkDescriptor]:
+    """Writes polygon chunk files based on source with polygons and chunk structure.
+
+    Args:
+        elevation_header: Information about cell grid projection to coordinate system.
+        chunk_size: Size of each chunk in cells.
+        polygon_masks: Source of polygons with associated mask values.
+        output_dir_path: Path to the directory where chunk files will be stored.
+        chunk_file_name_pattern: Format pattern used to generate chink file names.
+        support_mask_values: Optional indicator that masks should be added into output
+            as additional first column.
+
+    Returns:
+        List of descriptors for produced chunk files.
+    """
+    global_col_count = elevation_header.col_count
+    global_row_count = elevation_header.row_count
+    x_chunk_count = int((global_col_count + chunk_size - 1) / chunk_size)
+    y_chunk_count = int((global_row_count + chunk_size - 1) / chunk_size)
+
+    step = elevation_header.cell_size
+    chunk_bboxes: list[list[Tuple[float, float, float, float]]] = []
+    chunk_polygons: list[list[list[Tuple[geometry.Polygon, int]]]] = []
+    for y_chunk_index in range(0, y_chunk_count):
+        chunk_bboxes.append([])
+        chunk_polygons.append([])
+        for x_chunk_index in range(0, x_chunk_count):
+            bbox = _get_chunk_bounding_box(
+                elevation_header, chunk_size, y_chunk_index, x_chunk_index
+            )
+            chunk_bboxes[y_chunk_index].append(
+                (bbox[0] - step, bbox[1] - step, bbox[2] + step, bbox[3] + step)
+            )
+            chunk_polygons[y_chunk_index].append([])
+
+    for polygon_mask in polygon_masks:
```

Comment:
Do you know how much time this takes to run on files representative of the sizes we'll be dealing with?
Just asking because I see the nested loops, although I understand doing anything more clever would be quite a bit more effort and a premature optimization!

Reply:
It takes about 128 seconds on my local machine to load 1086217 polygons from buildings shape file. Next, 142 seconds to bin them into chunks in memory. And finally around a minute to store them all into polygon chunk text files.

Reply:
Great!